tensorflow-tools/src/main/java/org/tensorflow/tools/ndarray/impl/dense/AbstractDenseNdArray.java

```diff
@@ -110,6 +113,18 @@ protected AbstractDenseNdArray(DimensionalSpace dimensions) {
 
   abstract protected DataBuffer<T> buffer();
 
+  protected DataBuffer<T> trimBuffer() {
```

Comment:
This is disappointing, is there no way to make `equals` and `hashCode` just look at the view without having to actually perform an operation on them?

Reply:
We could avoid it by trimming the buffer every time we slice the array but this function is more on the critical performance path and I was hesitating changing it only to support equals/hashCode, which should be used less frequently. 

Trimming the buffer when the array is not segmented (i.e. caused by the usage of some of the `Index` classes) is quite easy to do but substantial changes need to be performed in the other case.

I'm not too happy with this last-minute discovery as well so would you be comfortable if we keep it like that and I'll add a TODO notice to take a deeper look at how to improve this aspect later?

Reply:
Yeah, I'm fine with it staying as is with a note to fix it later.

Reply:
Maybe we could add a flag on the NdArray noting it's not continuous in it's backing store, then have equals and hashcode fall back to something slower?

Reply:
Oh, we already have this flag (`dimensions.isSegmented()`), so I would now assume the buffer is trimmed if this flag is not set? I think it is better but still not perfect, I'll do this but still leave a note to review for later.

Reply:
Ok @Craigacp , I just pushed new substantial changes, this was required if I wanted to limit the impact on performances when slicing an array. So I added a new `DataBuffer.slice()` method, instead of using the double-allocating `DataBuffer.offset(i).narrow(s)` (running `NdArraBenchmark` show a significative improvement by doing this).